### PR TITLE
(#821) Allow for unbalanced JSON fragments in RSpec output

### LINKED ISSUE #821: Generated descriptions for certain test comparisons break JUnit output
Generated descriptions can cause the JSON parsing for JUnit output to fail.  When this occurs the created XML document will be empty.

### How to reproduce
Generate a new module with PDK.
```
$ pdk new module test-broken_junit
```

Create a unit test with the following content.
```
# ./spec/unit/broken_junit_spec.rb
# This test will break JUNIT output
RSpec.describe "broken-junit" do
  let(:large_nested_hash) do
    {
      single_one: 'single_one',
      single_two: 'single_two',
      single_three: 'single_three',
      single_four: 'single_four',
      single_five: 'single_five',
      nested: {
        nested_one: 'nested_one',
        nested_two: 'nested_two',
        nested_three: 'nested_three',
        nested_four: 'nested_four',
        nested_five: 'nested_five',
      },
    }
  end

  subject do
    large_nested_hash
  end

  it { is_expected.to be_a Hash }
  it { is_expected.to eq large_nested_hash }
end
```

Running the test should result in a success.
```
$ pdk test unit -v
pdk (INFO): Using Ruby 2.5.7
pdk (INFO): Using Puppet 6.11.0
[✔] Preparing to run the unit tests.
/opt/puppetlabs/pdk/private/ruby/2.5.7/bin/ruby -I/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/rspec-core-3.9.0/lib:/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/rspec-support-3.9.0/lib /opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/rspec-core-3.9.0/exe/rspec --pattern spec/\{aliases,classes,defines,functions,hosts,integration,plans,tasks,type_aliases,types,unit\}/\*\*/\*_spec.rb --format documentation

broken-junit
  is expected to be a kind of Hash
  is expected to eq {:nested=>{:nested_five=>"nested_five", :nested_four=>"nested_four", :nested_one=>"nested_one", :nest...>"single_four", :single_one=>"single_one", :single_three=>"single_three", :single_two=>"single_two"}

Finished in 0.00334 seconds (files took 0.17747 seconds to load)
2 examples, 0 failures
```

It is important to notice that the hash comparison test description output (`is expected to eq {:nested...`) has two opening curly braces (`{`) and only one closing curly brace (`}`).  The is an artifact of how the test descriptions are condensed for output.  This causes the processing for JUnit output to fail.

Running the test with the JUnit format specified should result in an empty document.
```
pdk (INFO): Using Ruby 2.5.7
pdk (INFO): Using Puppet 6.11.0
[✔] Preparing to run the unit tests.
[✔] Running unit tests.
<?xml version='1.0'?>
<testsuites/>
```

If you comment out enough of the `single_` lines to ensure that the output produces two closing curly braces (`}`) it will generate the XML as expected.

### Extra info
I dug through the `lib/pdk/tests/unit.rb` file and found that the output is processed by the `PDK::Util.find_first_json_in` method ([line 112](https://github.com/puppetlabs/pdk/blob/master/lib/pdk/tests/unit.rb#L112)).  I was able to confirm that the RSPEC command output JSON is valid.  So, I went ahead and chased that method into the `lib/pdk/util.rb` file.  [Line 219](https://github.com/puppetlabs/pdk/blob/master/lib/pdk/util.rb#L219) appears to be the culprit.  I was able to confirm that when that condition exists it does not properly capture the JSON output from the command.  Unfortunately, I don't have a fix to propose as I don't know the conditions that need to be satisfied with that REGEX.

### PR DESCRIPTION
**Pull request by scotje**
This is an initial attempt at making the PDK::Util.find_valid_json_in
method tolerate unbalanced JSON fragments nested as values inside of
a larger, valid JSON document.

The simple change of making the closing brace optional allows the regex to more greedily match the largest valid document, but I would like to see this tested against more complex samples before declaring this a sufficient fix.

Resolves #821

### CODE REVIEW

**Comment by scotje** [comment]
This isn't working for `--parallel` mode, going to iterate a bit more.

File: lib/pdk/util.rb [hunk 1]
```diff
@@ -196,7 +197,7 @@ def in_module_root?(path = Dir.pwd)
     # @return [Hash, nil] subset of text as Hash of first valid JSON found, or nil if no valid
     #   JSON found in the text
     def find_first_json_in(text)
-      find_valid_json_in(text)
+      PDK::Util::JSONFinder.new(text).objects.first
```
Comment (glennsarti) [inline lib/pdk/util.rb]: nit: Should this just be `find_all_json_in(text).first` ?

File: spec/acceptance/test_unit_spec.rb [hunk 2]
```diff
@@ -211,6 +211,77 @@
...
+
+              it { is_expected.to be_a Hash }
+              it { is_expected.to eq large_nested_hash }
+            end
+          EOF
+        end
+      end
+
+      after(:all) do
+        FileUtils.rm_rf(File.join('spec', 'unit'))
```
Comment (glennsarti) [inline spec/acceptance/test_unit_spec.rb]: This seems dangerous. What's cwd at this point?

File: lib/pdk/util.rb (hunk 1 above)
Comment (rodjek) [reply lib/pdk/util.rb]: :+1:

File: spec/acceptance/test_unit_spec.rb (hunk 2 above)
Comment (rodjek) [reply spec/acceptance/test_unit_spec.rb]: cwd is the test module directory, but i can replace this with a couple of safer methods

Comment (glennsarti) [reply spec/acceptance/test_unit_spec.rb]: 👍 I imagine my face if this ever went hay-wire would be 😕💥😢

Absolute path would be nice but that could be hard to get 🤷‍♂️

File: spec/acceptance/test_unit_spec.rb [hunk 3]
```diff
@@ -4,6 +4,16 @@
 describe 'pdk test unit', module_command: true do
   include_context 'with a fake TTY'
 
+  shared_context 'with spec file' do |filename, content|
+    around(:all) do |example|
+      path = File.join('spec', 'unit', filename)
+      FileUtils.mkdir_p(File.dirname(path))
+      File.open(path, 'w') { |f| f.puts content }
```
Comment (glennsarti) [inline spec/acceptance/test_unit_spec.rb]: Do we care about line endings here?  (not `wb` so on Windows it'll use CRLF)

Comment (rodjek) [reply spec/acceptance/test_unit_spec.rb]: Nope, for the purposes of these short lived files, the line endings don't matter

**Pull request merged**